### PR TITLE
Milestone 2: Ruby Projects

### PR DESCRIPTION
**Pull request by OdongAlican**
## Advanced Building Blocks Enumerable methods
This project was created using the following technologies.
- Ruby
- Rubocop
- Stickler

### CODE REVIEW

File: enumerables.rb [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    while index < size
+      if is_a? Array
+        yield self[index]
+      elsif is_a? Range
+        yield to_a[index]
+      end
+      index += 1
+    end
+  end
+
```
Comment (majovanilla) [inline enumerables.rb]: - [x] should have working `my_each` identical and returns the same thing as ruby's [`each`](https://ruby-doc.org/core-2.6.4/Array.html#method-i-each). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
Yours is not:
- Returning an enumerator when no block is given.

File: enumerables.rb [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+  def my_each_with_index
+    (0...size).each do |i|
+      yield(self[i], i)
+    end
+    self
+  end
```
Comment (majovanilla) [inline enumerables.rb]: - [x] should have working `my_each_with_index` identical and returns the same thing as ruby's [`each_with_index`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-each_with_index). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each_with_index.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

Yours is not Returning an enumerator when no block is given.

File: enumerables.rb [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+  def my_select
+    values = []
+    my_each do |val|
+      values.push(val) if yield(val)
+    end
+    values
+  end
```
Comment (majovanilla) [inline enumerables.rb]: - [x] should have working `my_select` identical and returns the same thing as ruby's [`select`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-select). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_select.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

Yours is not Returning an enumerator when no block is given.

File: enumerables.rb [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+  def my_all?
+    if block_given?
+      values = true
+      my_each { |val| values = false unless yield(val) }
+      values
+    else
+      empty?
+    end
+  end
```
Comment (majovanilla) [inline enumerables.rb]: - [x] should have working `my_all?` **(covered by automated tests)** identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

- Your method doesn't consider the possibility when no block or argument is given. It should return true if the enumerable contains truthy values and false if it contains at least one falsy value.

- Your method doesn't consider the possibility when a class is given as an argument `(some_enumerable.my_all?(String)).`

- Your method doesn't consider the possibility when a Regex is given as an argument `(some_enumerable.my_all?(/w/)).`

- Your method doesn't consider the possibility when a number is given as an argument `(some_enumerable.my_all?(5)).`

File: enumerables.rb [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+  def my_any?
+    if block_given?
+      values = false
+      my_each { |val| values = true if yield(val) }
+      values
+    else
+      true
+    end
+  end
```
Comment (majovanilla) [inline enumerables.rb]: - [x] should have working `my_any?` identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

- Your method doesn't consider the possibility when no block or argument is given. It should return true if the enumerable contains at least one truthy value.

- Your method doesn't consider the possibility when a class is given as an argument `(some_enumerable.my_any?(String)).`

- Your method doesn't consider the possibility when a Regex is given as an argument `(some_enumerable.my_any?(/w/)).`

- Your method doesn't consider the possibility when a string is given as an argument `(some_enumerable.my_any?('cow')).`

File: enumerables.rb [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+      values = true
+      my_each { |val| values = false if yield(val) }
+      values
+    elsif self[0].nil? && !include?(true)
+      return true
+    else
+      return false
+    end
+    values
+  end
```
Comment (majovanilla) [inline enumerables.rb]: - [x] should have working `my_none?` identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

- Your method doesn't consider the possibility when a class is given as an argument `(some_enumerable.my_none?(String)).`

- Your method doesn't consider the possibility when a Regex is given as an argument `(some_enumerable.my_none?(/w/)).`

- Your method doesn't consider the possibility when a number is given as an argument `(some_enumerable.my_none?(8)).`

File: enumerables.rb [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+
+  def my_count
+    count = 0
+    my_each do |num|
+      count += 1 if yield(num)
+    end
+    count
+  end
```
Comment (majovanilla) [inline enumerables.rb]: - [x] should have working `my_count` identical and returns the same thing as ruby's [`count`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-count). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_count.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

- Your method doesn't consider the possibility when no argument or block is given. It should return the count of all elements inside the enumerable.

- Your method doesn't consider the possibility when a number is given as an argument `(some_enumerable.my_count(5)).` It should return the count of all elements that match the given argument.

File: enumerables.rb [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+
+  def my_map
+    mapped = []
+    my_each do |i|
+      mapped << (proc.nil? ? proc.call(i) : yield(i))
+    end
+    mapped
+  end
```
Comment (majovanilla) [inline enumerables.rb]: - [x] should have working `my_map` **(covered by automated tests)** identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

- Your method should return an enumerator if no block or argument is given.

File: enumerables.rb [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+    each do |item|
+      updated_result << item
+    end
+    updated_result.delete_at(0)
+
+    updated_result.each do |val|
+      result = yield(result, val)
+    end
+    result
+  end
```
Comment (majovanilla) [inline enumerables.rb]: - [x] should have working `my_inject` identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

- Your method should work with symbols as arguments (:+) and perform the right operation.
- Your method should work with numbers and blocks as arguments `some_enumerable.my_inject(5) {block with some operation}` Should have the first number as the starting value of the result variable.

**Review by majovanilla — CHANGES_REQUESTED** [review]
Hey @OdongAlican , 👋

## Good job on this project! 👍

- Good Readme file
- Good Pull Request Message
- You are passing most of the test cases!

There are still some changes 👇 you need to implement to approve this request 🔍

- Check the commented code below.

## Please, make all REQUIRED changes and submit your request again! 😀

If you have any questions, feel free to send me a message through Slack (@majo)
I would be happy to help! 👩‍💻 


Happy coding 🤓

--Maria Reyes

**Comment by OdongAlican** [comment]
Made the requested changes

File: enumerables.rb [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+      my_each { |val| values = false if val != value }
+      values
+    elsif !block_given? && value.is_a?(String)
+      values = true
+      my_each { |val| values = false if val != value }
+      values
+    elsif !block_given? && value.is_a?(Regexp)
+      values = true
+      my_each { |val| values = false if val }
+      values
```
Comment (majovanilla) [inline enumerables.rb]: This part of your method is not working. When you check something like this: `['string', 'ring', 'spring'].my_all?(/ing/` should return true, but it's returning false.

File: enumerables.rb [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+    end
+  end
+
+  def my_any?(value = nil) # rubocop:disable Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
+    if block_given?
+      values = false
+      my_each { |val| values = true if yield(val) }
+      values
+    elsif !block_given? && value.is_a?(Regexp)
+      false
```
Comment (majovanilla) [inline enumerables.rb]: This part of your method is not working. When you check something like this: ['string', 'ring', 'spring'].my_all?(/ing/ should return true, but it's returning false.

File: enumerables.rb [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+    end
+  end
+
+  def my_none?(value = nil) # rubocop:disable Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
+    if block_given?
+      values = true
+      my_each { |val| values = false if yield(val) }
+      values
+    elsif !block_given? && value.is_a?(Regexp)
+      return true
```
Comment (majovanilla) [inline enumerables.rb]: This part of your method is not working. When you check something like this: ['string', 'ring', 'spring'].my_all?(/ing/ should return false, but it's returning true.

**Review by majovanilla — CHANGES_REQUESTED** [review]
Hey @OdongAlican , 👋

## You did a *great* job so far! 💜  You have a few things to cover before approving, but it's almost done!!

- Check the commented code below and implement the changes before submitting for a new review.

If you have any questions, feel free to send me a message through Slack (@majo)
I would be happy to help! 👩‍💻 


Happy coding 🤓

--Maria Reyes

**Comment by OdongAlican** [comment]
Thanks, @majovanilla for the feedback.  I updated the requested changes for Regex conditions in my_all, my_none and my_any methods. Please check it out!!

**Review by Torres-ssf — APPROVED** [review]
### Project Approved :tada: :+1: 
Hi there @OdongAlican, how are you?
Great job implementing the requested changes :smile: :nerd_face: 
**Your project is approved.**
Good luck with the next one.

**Sergio Torres**

**Pull request merged**